web/packages/teleterm/src/ui/ConnectMyComputer/DocumentConnectMyComputerStatus/DocumentConnectMyComputerStatus.tsx

```diff
@@ -198,91 +215,155 @@ function renderLabels(labelsList: tsh.Label[]): JSX.Element[] {
-        stackTrace: agentState.stackTrace,
-      };
-    }
-    case 'download-error': {
-      return {
-        Icon: StyledWarning,
-        title: 'Failed to download agent',
-        error: agentState.message,
-      };
-    }
-    case 'kill-error': {
-      return {
-        Icon: StyledWarning,
-        title: 'Failed to kill agent',
-        error: agentState.message,
-      };
-    }
-    case 'join-error': {
-      return {
-        Icon: StyledWarning,
-        title: 'Failed to join cluster',
-        error: agentState.message,
-      };
+          if (currentAction.agentProcessState.status === 'error') {
+            return {
+              Icon: StyledWarning,
+              title:
+                'Failed to start agent – an error occurred while spawning the agent process',
+              error: currentAction.agentProcessState.message,
+            };
+          }
+
+          if (currentAction.agentProcessState.status === 'exited') {
+            const { code, signal } = currentAction.agentProcessState;
+            return {
+              Icon: StyledWarning,
+              title: `Failed to start agent - the agent process quit unexpectedly with ${codeOrSignal(
+                code,
+                signal
+              )}`,
+              stackTrace: currentAction.agentProcessState.stackTrace,
+            };
+          }
+          break;
+        }
+        case 'success': {
+          return; // noop, not used, at this point it should be observe-process running.
+        }
+        default: {
+          return assertUnreachable(currentAction.attempt.status);
+        }
+      }
+      break;
     }
-    case 'process-error': {
-      return {
-        Icon: StyledWarning,
-        title: 'An error occurred to the agent process',
-        error: agentState.message,
-      };
+    case 'observe-process': {
+      switch (currentAction.agentProcessState.status) {
+        case 'not-started': {
+          return {
+            Icon: Moon,
+            title: 'Agent not running',
+          };
+        }
+        case 'running': {
+          return {
+            Icon: props => <CircleCheck {...props} color="success" />,
+            title: 'Agent running',
+          };
+        }
+        case 'exited': {
+          const { code, signal, exitedSuccessfully } =
+            currentAction.agentProcessState;
+
+          if (exitedSuccessfully) {
+            return {
+              Icon: Moon,
+              title: 'Agent not running',
+            };
+          } else {
+            return {
+              Icon: StyledWarning,
+              title: `Agent process exited with ${codeOrSignal(code, signal)}`,
+              stackTrace: currentAction.agentProcessState.stackTrace,
+            };
+          }
+        }
+        case 'error': {
+          // TODO(ravicious): This can probably happen only just before killing the process? Error
+          // during spawn is already handled by the 'start' branch.
```

Comment:
By that you mean that we can make the error message more specific?

Reply:
No, I think at the time I thought that if `error` was emitted from the child process during killing, then the `kill` branch here would handle it. But that's not the case because the `kill` attempt doesn't wait for process errors like the `start` attempt does.

It appears to me that we modeled `error`, and thus `start` and `kill` slightly wrong. I have to admit I didn't fully understand the concept of the `error` event, but after working on the cleanup daemon, I think I have a better grasp on it.

`error` shouldn't be a separate process state. Instead, `AgentRunner.start` should not resolve until either `spawn` is emitted or reject if `error` is emitted. Likewise, `kill` shouldn't resolve until `exit` is emitted or reject if `error` is emitted.

This would also simplify the code for starting the agent slightly, because we wouldn't need to listen to process errors (`throwOnAgentProcessErrors`), only the exit event.

Alas, I'm not sure if it's worth it to refactor this further right now, I'm going to just add comments about this.

Reply:
>error shouldn't be a separate process state. Instead, AgentRunner.start should not resolve until either spawn is emitted or reject if error is emitted. Likewise, kill shouldn't resolve until exit is emitted or reject if error is emitted.

This makes sense. 
I don't like this `error` event, the life would be so much easier if `spawn` or `kill` just throw this error.

>Alas, I'm not sure if it's worth it to refactor this further right now, I'm going to just add comments about this.

👍